File: java/knn/KNNBasico.java
/*
 * K-Nearest Neighbors (KNN) - Implementação Básica em Java
 * ========================================================
 * 
 * Algoritmo de classificação que determina a classe de um ponto baseado
 * na classe majoritária dos K vizinhos mais próximos.
 * 
 * Complexidade:
 * - Tempo: O(n * d) para cada predição
 * - Espaço: O(n * d) para armazenar dados de treino
 * 
 * Características desta implementação:
 * - Orientação a objetos clara
 * - Encapsulamento adequado
 * - Comentários extensos para aprendizado
 * - Tratamento de exceções
 */

import java.util.*;

/**
 * Representa um ponto no espaço n-dimensional com sua classe
 */
class Ponto {
    private double[] coordenadas;
    private String classe;
    
    public Ponto(double[] coordenadas, String classe) {
        this.coordenadas = Arrays.copyOf(coordenadas, coordenadas.length);
        this.classe = classe;
    }
    
    public double[] getCoordenadas() {
        return Arrays.copyOf(coordenadas, coordenadas.length);
    }
    
    public String getClasse() {
        return classe;
    }
    
    @Override
    public String toString() {
        return String.format("Ponto{coordenadas=%s, classe='%s'}", 
                           Arrays.toString(coordenadas), classe);
    }
}

/**
 * Representa um vizinho com sua distância para o ponto de consulta
 */
class Vizinho implements Comparable<Vizinho> {
    private Ponto ponto;
    private double distancia;
    
    public Vizinho(Ponto ponto, double distancia) {
        this.ponto = ponto;
        this.distancia = distancia;
    }
    
    public Ponto getPonto() {
        return ponto;
    }
    
    public double getDistancia() {
        return distancia;
    }
    
    @Override
    public int compareTo(Vizinho outro) {
        return Double.compare(this.distancia, outro.distancia);
    }
    
    @Override
    public String toString() {
        return String.format("Vizinho{distância=%.3f, classe='%s'}", 
                           distancia, ponto.getClasse());
    }
}

/**
 * Implementação básica do algoritmo K-Nearest Neighbors
 */
public class KNNBasico {
    private int k;
    private List<Ponto> dadosTreino;
    
    /**
     * Construtor do classificador KNN
     * @param k Número de vizinhos mais próximos a considerar
     */
    public KNNBasico(int k) {
        if (k <= 0) {
            throw new IllegalArgumentException("K deve ser positivo");
        }
        
        this.k = k;
        this.dadosTreino = new ArrayList<>();
        
        System.out.printf("🎯 KNN Básico inicializado com k=%d%n", k);
    }
    
    /**
     * Calcula a distância euclidiana entre dois pontos
     * @param p1 Primeiro ponto
     * @param p2 Segundo ponto
     * @return Distância euclidiana
     * @throws IllegalArgumentException se os pontos têm dimensões diferentes
     */
    public double calcularDistanciaEuclidiana(double[] p1, double[] p2) {
        if (p1.length != p2.length) {
            throw new IllegalArgumentException("Pontos devem ter a mesma dimensão");
        }
        
        System.out.println("   📏 Calculando distância euclidiana:");
        
        double somaQuadrados = 0.0;
        for (int i = 0; i < p1.length; i++) {
            double diferenca = p1[i] - p2[i];
            double quadrado = diferenca * diferenca;
            somaQuadrados += quadrado;
            
            System.out.printf("      Dimensão %d: (%.3f - %.3f)² = %.3f%n", 
                            i + 1, p1[i], p2[i], quadrado);
        }
        
        double distancia = Math.sqrt(somaQuadrados);
        System.out.printf("      Distância: √%.3f = %.3f%n", somaQuadrados, distancia);
        
        return distancia;
    }
    
    /**
     * Treina o modelo KNN (armazena os dados de treinamento)
     * @param dados Lista de pontos para treinamento
     */
    public void treinar(List<Ponto> dados) {
        this.dadosTreino = new ArrayList<>(dados);
        
        System.out.printf("📚 Modelo treinado com %d exemplos%n", dadosTreino.size());
        
        // Conta distribuição das classes
        Map<String, Integer> contadorClasses = new HashMap<>();
        for (Ponto ponto : dados) {
            contadorClasses.merge(ponto.getClasse(), 1, Integer::sum);
        }
        
        System.out.print("📊 Distribuição das classes: ");
        contadorClasses.forEach((classe, count) -> 
            System.out.printf("%s:%d ", classe, count));
        System.out.println();
    }
    
    /**
     * Encontra os k vizinhos mais próximos de um ponto
     * @param pontoConsulta Ponto para o qual encontrar vizinhos
     * @return Lista dos k vizinhos mais próximos
     * @throws IllegalStateException se o modelo não foi treinado
     */
    public List<Vizinho> encontrarKVizinhos(double[] pontoConsulta) {
        if (dadosTreino.isEmpty()) {
            throw new IllegalStateException("Modelo não foi treinado. Chame treinar() primeiro.");
        }
        
        System.out.printf("%n🔍 Buscando %d vizinhos mais próximos para %s%n", 
                         k, Arrays.toString(pontoConsulta));
        
        // Calcula distâncias para todos os pontos de treino
        List<Vizinho> vizinhos = new ArrayList<>();
        
        for (int i = 0; i < dadosTreino.size(); i++) {
            Ponto pontoTreino = dadosTreino.get(i);
            
            System.out.printf("%n📏 Calculando distância para ponto %d: %s (classe: %s)%n", 
                            i + 1, Arrays.toString(pontoTreino.getCoordenadas()), 
                            pontoTreino.getClasse());
            
            double distancia = calcularDistanciaEuclidiana(pontoConsulta, 
                                                         pontoTreino.getCoordenadas());
            
            vizinhos.add(new Vizinho(pontoTreino, distancia));
        }
        
        // Ordena vizinhos por distância (crescente)
        Collections.sort(vizinhos);
        
        // Retorna apenas os k primeiros
        List<Vizinho> kVizinhos = vizinhos.subList(0, Math.min(k, vizinhos.size()));
        
        System.out.printf("%n🎯 %d vizinhos mais próximos:%n", kVizinhos.size());
        for (int i = 0; i < kVizinhos.size(); i++) {
            Vizinho vizinho = kVizinhos.get(i);
            System.out.printf("   %dº vizinho: distância=%.3f, classe=%s%n", 
                            i + 1, vizinho.getDistancia(), vizinho.getPonto().getClasse());
        }
        
        return kVizinhos;
    }
    
    /**
     * Classifica um ponto baseado nos k vizinhos mais próximos
     * @param pontoConsulta Ponto a ser classificado
     * @return Classe prevista
     */
    public String classificar(double[] pontoConsulta) {
        // Encontra os k vizinhos mais próximos
        List<Vizinho> kVizinhos = encontrarKVizinhos(pontoConsulta);
        
        // Conta a frequência de cada classe
        Map<String, Integer> contadorClasses = new HashMap<>();
        for (Vizinho vizinho : kVizinhos) {
            String classe = vizinho.getPonto().getClasse();
            contadorClasses.merge(classe, 1, Integer::sum);
        }
        
        System.out.print("%n📊 Contagem de classes nos vizinhos: ");
        contadorClasses.forEach((classe, count) -> 
            System.out.printf("%s:%d ", classe, count));
        System.out.println();
        
        // Encontra a classe mais frequente
        String classePrevista = contadorClasses.entrySet().stream()
            .max(Map.Entry.comparingByValue())
            .map(Map.Entry::getKey)
            .orElse("desconhecida");
        
        int maxCount = contadorClasses.get(classePrevista);
        double confianca = (double) maxCount / kVizinhos.size();
        
        System.out.printf("🎯 Classe prevista: %s%n", classePrevista);
        System.out.printf("🔢 Confiança: %.1f%% (%d/%d vizinhos)%n", 
                         confianca * 100, maxCount, kVizinhos.size());
        
        return classePrevista;
    }
    
    /**
     * Demonstração do algoritmo KNN com exemplo prático
     */
    public static void demonstracaoKNNBasico() {
        System.out.println("=".repeat(60));
        System.out.println("🤖 DEMONSTRAÇÃO: K-Nearest Neighbors (KNN) - Versão Básica");
        System.out.println("=".repeat(60));
        
        // Cria dados de exemplo: classificação de animais
        // [peso_kg, altura_cm] -> tipo_animal
        List<Ponto> dadosAnimais = Arrays.asList(
            new Ponto(new double[]{0.5, 20}, "gato"),
            new Ponto(new double[]{0.7, 25}, "gato"),
            new Ponto(new double[]{0.3, 18}, "gato"),
            new Ponto(new double[]{25, 60}, "cão"),
            new Ponto(new double[]{30, 65}, "cão"),
            new Ponto(new double[]{20, 55}, "cão"),
            new Ponto(new double[]{500, 180}, "cavalo"),
            new Ponto(new double[]{450, 175}, "cavalo"),
            new Ponto(new double[]{520, 185}, "cavalo")
        );
        
        System.out.printf("🐾 Dataset: Classificação de animais com %d exemplos%n", 
                         dadosAnimais.size());
        System.out.println("📊 Características: [peso_kg, altura_cm]");
        
        // Inicializa o classificador
        KNNBasico knn = new KNNBasico(3);
        
        // Treina o modelo
        knn.treinar(dadosAnimais);
        
        // Testa com diferentes pontos
        double[][] pontosTeste = {
            {0.6, 22},   // Deve ser gato
            {28, 62},    // Deve ser cão
            {480, 178},  // Deve ser cavalo
            {10, 40}     // Caso ambíguo
        };
        
        System.out.println("\n" + "=".repeat(40));
        System.out.println("🧪 TESTANDO CLASSIFICAÇÕES");
        System.out.println("=".repeat(40));
        
        for (int i = 0; i < pontosTeste.length; i++) {
            double[] ponto = pontosTeste[i];
            System.out.printf("%n🔍 TESTE %d: Classificando ponto %s%n", 
                            i + 1, Arrays.toString(ponto));
            System.out.println("-".repeat(40));
            
            try {
                String classePrevista = knn.classificar(ponto);
                System.out.printf("✅ Resultado final: %s → %s%n", 
                                Arrays.toString(ponto), classePrevista);
            } catch (Exception e) {
                System.out.printf("❌ Erro: %s%n", e.getMessage());
            }
        }
    }
    
    /**
     * Demonstra como diferentes valores de k afetam a classificação
     */
    public static void exemploComparacaoK() {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("📈 COMPARAÇÃO: Efeito de diferentes valores de K");
        System.out.println("=".repeat(60));
        
        // Dados simples para demonstração
        List<Ponto> dados = Arrays.asList(
            new Ponto(new double[]{1, 1}, "A"),
            new Ponto(new double[]{1, 2}, "A"),
            new Ponto(new double[]{2, 1}, "A"),
            new Ponto(new double[]{5, 5}, "B"),
            new Ponto(new double[]{5, 6}, "B"),
            new Ponto(new double[]{6, 5}, "B"),
            new Ponto(new double[]{3, 3}, "A") // Ponto intermediário
        );
        
        double[] pontoTeste = {3, 4};
        int[] valoresK = {1, 3, 5};
        
        for (int k : valoresK) {
            System.out.printf("%n🎯 Testando com k=%d%n", k);
            System.out.println("-".repeat(30));
            
            try {
                KNNBasico knn = new KNNBasico(k);
                knn.treinar(dados);
                
                String resultado = knn.classificar(pontoTeste);
                System.out.printf("📊 Com k=%d: %s → %s%n", 
                                k, Arrays.toString(pontoTeste), resultado);
            } catch (Exception e) {
                System.out.printf("❌ Erro: %s%n", e.getMessage());
            }
        }
    }
    
    /**
     * Sugere exercícios práticos para aprendizado
     */
    public static void exerciciosPraticos() {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("🎓 EXERCÍCIOS SUGERIDOS:");
        System.out.println("=".repeat(60));
        System.out.println("1. Implemente outras métricas de distância (Manhattan, Minkowski)");
        System.out.println("2. Adicione validação mais robusta dos dados de entrada");
        System.out.println("3. Implemente KNN para regressão (predizer valores numéricos)");
        System.out.println("4. Adicione função para encontrar o melhor k automaticamente");
        System.out.println("5. Crie classes para diferentes métricas de avaliação");
        System.out.println("6. Implemente normalização automática dos dados");
        System.out.println("7. Adicione suporte para pesos baseados na distância");
        System.out.println("8. Crie interface gráfica para visualização");
    }
    
    /**
     * Método principal - executa todas as demonstrações
     */
    public static void main(String[] args) {
        // Executa as demonstrações
        demonstracaoKNNBasico();
        exemploComparacaoK();
        exerciciosPraticos();
    }
}

File: go/knn/knn_basico.go
/*
K-Nearest Neighbors (KNN) - Implementação Básica em Go
======================================================

Algoritmo de classificação que determina a classe de um ponto baseado
na classe majoritária dos K vizinhos mais próximos.

Complexidade:
- Tempo: O(n * d) para cada predição
- Espaço: O(n * d) para armazenar dados de treino

Características desta implementação:
- Foco na clareza e entendimento
- Comentários extensos
- Estruturas simples
- Visualização do processo
*/

package main

import (
	"fmt"
	"math"
	"sort"
	"strings"
)

// Ponto representa um ponto no espaço n-dimensional
type Ponto struct {
	Coordenadas []float64 // Características do ponto
	Classe      string    // Classe/label do ponto
}

// String retorna representação textual do ponto
func (p Ponto) String() string {
	return fmt.Sprintf("Ponto{Coord: %v, Classe: %s}", p.Coordenadas, p.Classe)
}

// Vizinho representa um vizinho com sua distância
type Vizinho struct {
	Ponto     Ponto
	Distancia float64
}

// KNNBasico implementa o algoritmo K-Nearest Neighbors básico
type KNNBasico struct {
	K           int     // Número de vizinhos a considerar
	DadosTreino []Ponto // Dados de treinamento
}

// NovoKNNBasico cria uma nova instância do classificador KNN
func NovoKNNBasico(k int) *KNNBasico {
	fmt.Printf("🎯 KNN Básico inicializado com k=%d\n", k)
	return &KNNBasico{
		K:           k,
		DadosTreino: make([]Ponto, 0),
	}
}

// CalcularDistanciaEuclidiana calcula a distância euclidiana entre dois pontos
func (knn *KNNBasico) CalcularDistanciaEuclidiana(p1, p2 []float64) (float64, error) {
	if len(p1) != len(p2) {
		return 0, fmt.Errorf("pontos devem ter a mesma dimensão")
	}

	fmt.Printf("   📏 Calculando distância euclidiana:\n")

	somaQuadrados := 0.0
	for i := 0; i < len(p1); i++ {
		diferenca := p1[i] - p2[i]
		quadrado := diferenca * diferenca
		somaQuadrados += quadrado
		fmt.Printf("      Dimensão %d: (%.3f - %.3f)² = %.3f\n",
			i+1, p1[i], p2[i], quadrado)
	}

	distancia := math.Sqrt(somaQuadrados)
	fmt.Printf("      Distância: √%.3f = %.3f\n", somaQuadrados, distancia)

	return distancia, nil
}

// Treinar armazena os dados de treinamento
func (knn *KNNBasico) Treinar(dados []Ponto) {
	knn.DadosTreino = make([]Ponto, len(dados))
	copy(knn.DadosTreino, dados)

	fmt.Printf("📚 Modelo treinado com %d exemplos\n", len(knn.DadosTreino))

	// Conta distribuição das classes
	contadorClasses := make(map[string]int)
	for _, ponto := range dados {
		contadorClasses[ponto.Classe]++
	}

	fmt.Printf("📊 Distribuição das classes: ")
	for classe, count := range contadorClasses {
		fmt.Printf("%s:%d ", classe, count)
	}
	fmt.Println()
}

// EncontrarKVizinhos encontra os k vizinhos mais próximos de um ponto
func (knn *KNNBasico) EncontrarKVizinhos(pontoConsulta []float64) ([]Vizinho, error) {
	if len(knn.DadosTreino) == 0 {
		return nil, fmt.Errorf("modelo não foi treinado")
	}

	fmt.Printf("\n🔍 Buscando %d vizinhos mais próximos para %v\n",
		knn.K, pontoConsulta)

	// Calcula distâncias para todos os pontos de treino
	var vizinhos []Vizinho

	for i, pontoTreino := range knn.DadosTreino {
		fmt.Printf("\n📏 Calculando distância para ponto %d: %v (classe: %s)\n",
			i+1, pontoTreino.Coordenadas, pontoTreino.Classe)

		distancia, err := knn.CalcularDistanciaEuclidiana(pontoConsulta, pontoTreino.Coordenadas)
		if err != nil {
			return nil, err
		}

		vizinhos = append(vizinhos, Vizinho{
			Ponto:     pontoTreino,
			Distancia: distancia,
		})
	}

	// Ordena vizinhos por distância (crescente)
	sort.Slice(vizinhos, func(i, j int) bool {
		return vizinhos[i].Distancia < vizinhos[j].Distancia
	})

	// Retorna apenas os k primeiros
	kVizinhos := vizinhos
	if len(vizinhos) > knn.K {
		kVizinhos = vizinhos[:knn.K]
	}

	fmt.Printf("\n🎯 %d vizinhos mais próximos:\n", len(kVizinhos))
	for i, vizinho := range kVizinhos {
		fmt.Printf("   %dº vizinho: distância=%.3f, classe=%s\n",
			i+1, vizinho.Distancia, vizinho.Ponto.Classe)
	}

	return kVizinhos, nil
}

// Classificar classifica um ponto baseado nos k vizinhos mais próximos
func (knn *KNNBasico) Classificar(pontoConsulta []float64) (string, error) {
	// Encontra os k vizinhos mais próximos
	kVizinhos, err := knn.EncontrarKVizinhos(pontoConsulta)
	if err != nil {
		return "", err
	}

	// Conta a frequência de cada classe
	contadorClasses := make(map[string]int)
	for _, vizinho := range kVizinhos {
		contadorClasses[vizinho.Ponto.Classe]++
	}

	fmt.Printf("\n📊 Contagem de classes nos vizinhos: ")
	for classe, count := range contadorClasses {
		fmt.Printf("%s:%d ", classe, count)
	}
	fmt.Println()

	// Encontra a classe mais frequente
	classePrevista := ""
	maxCount := 0

	for classe, count := range contadorClasses {
		if count > maxCount {
			maxCount = count
			classePrevista = classe
		}
	}

	confianca := float64(maxCount) / float64(len(kVizinhos))

	fmt.Printf("🎯 Classe prevista: %s\n", classePrevista)
	fmt.Printf("🔢 Confiança: %.1f%% (%d/%d vizinhos)\n",
		confianca*100, maxCount, len(kVizinhos))

	return classePrevista, nil
}

// demonstracaoKNNBasico executa uma demonstração completa do algoritmo
func demonstracaoKNNBasico() {
	fmt.Println("=" + strings.Repeat("=", 59))
	fmt.Println("🤖 DEMONSTRAÇÃO: K-Nearest Neighbors (KNN) - Versão Básica")
	fmt.Println("=" + strings.Repeat("=", 59))

	// Cria dados de exemplo: classificação de frutas
	// [doçura, acidez] -> tipo_fruta
	dadosFrutas := []Ponto{
		{Coordenadas: []float64{8.0, 2.0}, Classe: "maçã"},
		{Coordenadas: []float64{7.5, 2.5}, Classe: "maçã"},
		{Coordenadas: []float64{8.2, 1.8}, Classe: "maçã"},
		{Coordenadas: []float64{3.0, 8.0}, Classe: "limão"},
		{Coordenadas: []float64{3.5, 7.5}, Classe: "limão"},
		{Coordenadas: []float64{2.8, 8.2}, Classe: "limão"},
		{Coordenadas: []float64{9.0, 3.0}, Classe: "banana"},
		{Coordenadas: []float64{8.8, 3.5}, Classe: "banana"},
		{Coordenadas: []float64{9.2, 2.8}, Classe: "banana"},
	}

	fmt.Printf("🍎 Dataset: Classificação de frutas com %d exemplos\n", len(dadosFrutas))
	fmt.Println("📊 Características: [doçura, acidez]")

	// Inicializa o classificador
	knn := NovoKNNBasico(3)

	// Treina o modelo
	knn.Treinar(dadosFrutas)

	// Testa com diferentes pontos
	pontosTeste := [][]float64{
		{7.8, 2.2}, // Deve ser maçã
		{3.2, 7.8}, // Deve ser limão
		{9.1, 3.2}, // Deve ser banana
		{6.0, 5.0}, // Caso ambíguo
	}

	fmt.Println("\n" + strings.Repeat("=", 40))
	fmt.Println("🧪 TESTANDO CLASSIFICAÇÕES")
	fmt.Println(strings.Repeat("=", 40))

	for i, ponto := range pontosTeste {
		fmt.Printf("\n🔍 TESTE %d: Classificando ponto %v\n", i+1, ponto)
		fmt.Println(strings.Repeat("-", 40))

		classePrevista, err := knn.Classificar(ponto)
		if err != nil {
			fmt.Printf("❌ Erro: %v\n", err)
			continue
		}

		fmt.Printf("✅ Resultado final: %v → %s\n", ponto, classePrevista)
	}
}

// exemploComparacaoK demonstra como diferentes valores de k afetam o resultado
func exemploComparacaoK() {
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("📈 COMPARAÇÃO: Efeito de diferentes valores de K")
	fmt.Println(strings.Repeat("=", 60))

	// Dados simples para demonstração
	dados := []Ponto{
		{Coordenadas: []float64{1, 1}, Classe: "A"},
		{Coordenadas: []float64{1, 2}, Classe: "A"},
		{Coordenadas: []float64{2, 1}, Classe: "A"},
		{Coordenadas: []float64{5, 5}, Classe: "B"},
		{Coordenadas: []float64{5, 6}, Classe: "B"},
		{Coordenadas: []float64{6, 5}, Classe: "B"},
		{Coordenadas: []float64{3, 3}, Classe: "A"}, // Ponto intermediário
	}

	pontoTeste := []float64{3, 4}

	valoresK := []int{1, 3, 5}

	for _, k := range valoresK {
		fmt.Printf("\n🎯 Testando com k=%d\n", k)
		fmt.Println(strings.Repeat("-", 30))

		knn := NovoKNNBasico(k)
		knn.Treinar(dados)

		resultado, err := knn.Classificar(pontoTeste)
		if err != nil {
			fmt.Printf("❌ Erro: %v\n", err)
			continue
		}

		fmt.Printf("📊 Com k=%d: %v → %s\n", k, pontoTeste, resultado)
	}
}

// exerciciosPraticos sugere exercícios para prática
func exerciciosPraticos() {
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("🎓 EXERCÍCIOS SUGERIDOS:")
	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("1. Implemente outras métricas de distância (Manhattan, Minkowski)")
	fmt.Println("2. Adicione validação dos dados de entrada")
	fmt.Println("3. Implemente KNN para regressão (predizer valores numéricos)")
	fmt.Println("4. Adicione função para encontrar o melhor k automaticamente")
	fmt.Println("5. Crie visualização dos resultados (se usar biblioteca gráfica)")
	fmt.Println("6. Implemente normalização automática dos dados")
	fmt.Println("7. Adicione suporte para pesos baseados na distância")
}

func main() {
	// Executa as demonstrações
	demonstracaoKNNBasico()
	exemploComparacaoK()
	exerciciosPraticos()
}
